Cache and base uri update

sethwoodworth opened the pull request:
This PR is the collaboration between myself and @becomesaflame from PR #11 . 
Also updates the `index.json` url and fixes #18 
Although this may duplicate the functionality of PR #8, which includes many other improvements.

sethwoodworth committed:
Caches all tldr-pages and creates cache on 1st run

shellcheck linting
whitespace removal

Jarnolandia commented on tldr:
```diff
@@ -168,12 +172,23 @@ path_for_cmd() {
+cache_tldr() {
+    curl -sfL https://github.com/tldr-pages/tldr/tarball/master > $configdir/tldr-pages.tar.gz
+    tar xzf "$configdir"/tldr-pages.tar.gz -C "$configdir"
+    mv -f "$configdir"/tldr-pages-tldr-*/pages/* "$configdir"
```
Add --backup to mv so mv doesn't cry that mv: cannot move source to destination: Directory not empty.

Jarnolandia replied:
And what comes to latter, it would be better to do

>for dir in "$configdir"/tldr-pages-tldr-*/pages/*/; do
>	dir=${dir%*/}
>	echo ${dir##*/}
>	mv -f "$dir" "$configdir"
>done

With that there is no copy behavior with --backup nor mv cry Directory not empty.

becomesaflame replied:
I'm happy to make an update here. Can you help me understand what the problem with the current `mv -f *` behavior is?  `mv -f` should overwrite a directory.

sethwoodworth replied:
Oops, it looks like @ghost is a now-deleted user.  @raylee Do you understand this comment?

raylee replied:
@sethwoodworth It would happen in a situation like this:
$ cd /tmp
$ mkdir -p a/1/1 b/1/1
$ mv -f b/* a
mv: rename b/1 to a/1: Directory not empty

mv won't overwrite a destination direction which contains files (correctly), as the user likely didn't want that.

raylee replied:
Working backward on the comments, I'm not sure `dir=${dir%*/}` does what's intended. The % here matches the shortest regex so the * is superfluous, and the expression will remove only a trailing slash if it exists:
```
$ dir=a/b/c/; echo ${dir%/} ${dir%*/}
a/b/c a/b/c
```

becomesaflame replied:
Ah, I see! I didn't realize `mv -f` behaved that way.  Thank you for the example. 

Perhaps it would be best to do something like this:
```
for dir in "$configdir"/tldr-pages-tldr-*/pages/*/; do
    dir=${dir%*/}
    rm -rf "$configdir/$dir"
    mv "$dir" "$configdir"
done
```
That way we don't have to worry about any further nested folders, or about backups already existing.

raylee replied:
Okay, I've walked through the above code by hand with the current tldr pages and I see what the OP is trying to do now. But I think there's an easier way.

Both GNU and BSD tar support a `--strip-components=n` option, which removes the first `n` path components from the filenames during extraction. So the above can be simplified to:

```
curl -sfL https://github.com/tldr-pages/tldr/tarball/master > "$configdir"/tldr-pages.tar.gz
tar xzf "$configdir"/tldr-pages.tar.gz --strip-components=2 -C "$configdir"
```
the `mv` can then be removed, and there'll be no need for the for loop suggestion above.

Sound good?

becomesaflame replied:
Great solution - thanks!

raylee commented on tldr:
```diff
@@ -24,6 +23,11 @@ config() {
+        cache_tldr
+    fi
+
     [ -d "$configdir" ] || mkdir -p "$configdir"
```
Right, this sorta duplicates the next line:

 [ -d "$configdir" ] || mkdir -p "$configdir"

Which says test if $configdir exists, if not create it. For people who have an existing tldr install, we'll want to gracefully upgrade them with the cached files, yes? In which case we probably want to add a second test below that, something like:

recent $configdir/tldr-pages.tar.gz || cache_tldr

which would let us remove the new if/fi block introduced above.

raylee commented on tldr:
```diff
@@ -184,14 +199,15 @@ get_tldr() {
     -l, --list:      show all available pages
     -p, --platform:  show page from specific platform rather than autodetecting
     -u, --update:    update, force retrieving latest copies of locally cached files
+    -c, --cache:     cache a local copy of all tldr commands
```
Suggested change for the text: "update local cache of all tldr commands"

becomesaflame committed:
Incorporated suggested changes to pull request

Deleted redundant line that creates config dir
Rephrased -cache flag description

becomesaflame commented:
Thanks for the feedback, @raylee and @ghost.  We've addressed the comments and pushed changes and are ready to implement any other suggestions you might have.  We should have some more time now to work on this - sorry for the long delay.

raylee committed:
Merge branch 'master' into cache-and-base-uri-update

raylee commented:
Merged with a few changes. Thanks everyone!
